server/s3_copy_object_action.cc

```diff
@@ -82,6 +82,158 @@ void S3CopyObjectAction::setup_steps() {
+
+  if ((source_object_list_oid.u_hi == 0ULL &&
+       source_object_list_oid.u_lo == 0ULL) ||
+      (source_object_version_list_oid.u_hi == 0ULL &&
+       source_object_version_list_oid.u_lo == 0ULL)) {
+    // Object list index and version list index missing.
+    fetch_source_object_info_failed();
+  } else {
+    source_object_metadata =
+        object_metadata_factory->create_object_metadata_obj(
+            request, source_object_list_oid, source_bucket_name,
+            source_object_name);
+
+    source_object_metadata->set_objects_version_list_index_oid(
+        source_object_version_list_oid);
+
+    source_object_metadata->load(
+        std::bind(&S3CopyObjectAction::fetch_source_object_info_success, this),
+        std::bind(&S3CopyObjectAction::fetch_source_object_info_failed, this));
+  }
+  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+}
+
+void S3CopyObjectAction::fetch_source_object_info_success() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  s3_log(S3_LOG_DEBUG, request_id,
+         "Successfully fetched source object metadata\n");
+
+  if (MaxCopyObjectSourceSize < source_object_metadata->get_content_length()) {
+    s3_copy_action_state = S3CopyObjectActionState::validationFailed;
+    set_s3_error("InvalidRequest");
+    send_response_to_s3_client();
+  } else {
+    next();
+  }
+  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+}
+
+void S3CopyObjectAction::fetch_source_object_info_failed() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+
+  s3_copy_action_state = S3CopyObjectActionState::validationFailed;
+
+  m0_uint128 source_object_list_oid =
+      source_bucket_metadata->get_object_list_index_oid();
+  if (source_object_list_oid.u_hi == 0ULL &&
+      source_object_list_oid.u_lo == 0ULL) {
+    s3_log(S3_LOG_ERROR, request_id, "Object not found\n");
+    set_s3_error("NoSuchKey");
+  } else {
+    if (S3ObjectMetadataState::missing == source_object_metadata->get_state()) {
+      set_s3_error("NoSuchKey");
+    } else if (S3ObjectMetadataState::failed_to_launch ==
+               source_object_metadata->get_state()) {
+      s3_log(S3_LOG_ERROR, request_id,
+             "Source object metadata load operation failed due to pre launch "
+             "failure\n");
+      set_s3_error("ServiceUnavailable");
+    } else {
+      s3_log(S3_LOG_DEBUG, request_id, "Source object metadata fetch failed\n");
+      set_s3_error("InternalError");
+    }
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+}
+
+// Validate source bucket and object
+void S3CopyObjectAction::validate_copyobject_request() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  get_source_bucket_and_object();
+
+  if (source_bucket_name.empty() || source_object_name.empty()) {
+    set_s3_error("InvalidArgument");
+    send_response_to_s3_client();
+  } else if (if_source_and_destination_same()) {
+    s3_copy_action_state = S3CopyObjectActionState::validationFailed;
+    set_s3_error("InvalidRequest");
```

Comment:
As per current s3_error_messages.json file
  "InvalidRequest": {
    "Description": "Specifying both Canned ACLs and Header Grants is not allowed",
    "httpcode": 400
  },
The Description seems to be incorrect there

Reply:
I have handled this in the  S3CopyObjectAction class send_response_to_s3_client().

There I set the error message according to the CopyObject action error.